contrib/gcppubsub/pkg/controller/channel/reconcile.go

```diff
@@ -169,100 +144,70 @@ func (r *reconciler) reconcile(ctx context.Context, c *eventingv1alpha1.Channel)
 
 	// First we will plan all the names out for steps 3 and 4 persist them to status.internal. Then, on a
 	// subsequent reconcile, we manipulate all the GCP resources in steps 3 and 4.
-
 	originalPCS, err := pubsubutil.GetInternalStatus(c)
 	if err != nil {
 		logging.FromContext(ctx).Error("Unable to read the status.internal", zap.Error(err))
-		r.recorder.Eventf(c, v1.EventTypeWarning, channelReadStatusFailed, "Failed to read Channel's status.internal: %v", err)
-		return false, err
+		recorder.Eventf(c, v1.EventTypeWarning, channelReadStatusFailed, "Failed to read Channel's status.internal: %v", err)
+		return false, reconcile.Result{}, err
 	}
 
 	// Regardless of what we are going to do, we need GCP credentials to do it.
 	gcpCreds, err := pubsubutil.GetCredentials(ctx, r.client, r.defaultSecret, r.defaultSecretKey)
 	if err != nil {
 		logging.FromContext(ctx).Info("Unable to generate GCP creds", zap.Error(err))
-		r.recorder.Eventf(c, v1.EventTypeWarning, gcpCredentialsReadFailed, "Failed to generate GCP credentials: %v", err)
-		return false, err
-	}
-
-	if c.DeletionTimestamp != nil {
-		// K8s garbage collection will delete the K8s service and VirtualService for this channel.
-		// All the subs should be deleted.
-		subsToSync := &syncSubs{
-			subsToDelete: originalPCS.Subscriptions,
-		}
-		// Topic is nil because it is only used for sub creation, not deletion.
-		err = r.syncSubscriptions(ctx, originalPCS, gcpCreds, nil, subsToSync)
-		if err != nil {
-			r.recorder.Eventf(c, v1.EventTypeWarning, subscriptionSyncFailed, "Failed to sync Subscription for the Channel: %v", err)
-			return false, err
-		}
-		err = r.deleteTopic(ctx, originalPCS, gcpCreds)
-		if err != nil {
-			r.recorder.Eventf(c, v1.EventTypeWarning, topicDeleteFailed, "Failed to delete Topic for the Channel: %v", err)
-			return false, err
-		}
-		util.RemoveFinalizer(c, finalizerName)
-		return false, nil
-	}
-
-	// If we are adding the finalizer for the first time, then ensure that finalizer is persisted
-	// before manipulating GCP PubSub, which will not be automatically garbage collected by K8s if
-	// this Channel is deleted.
-	if addFinalizerResult := util.AddFinalizer(c, finalizerName); addFinalizerResult == util.FinalizerAdded {
```

Comment:
So one thing that we lose is that with the existing code, the finalizer is only added if a credential and a PubSub client can be created. If for whatever reason they can't (e.g. the secret is badly formed), then the finalizer is not added, so the `Channel` can be deleted easily (without manually removing the finalizer).

We could either accept that we no longer have this nicety. Or we could do something else to indicate we haven't manipulated the real world at all (in fact I think the code is already setup to do this easily).